### PR TITLE
fix(security): never free PMIx-owned memory in nb credential/validation callbacks; deep-copy instead

### LINKED ISSUE #437: critical: security.rs nb callbacks free memory PMIx owns/frees — double-free and free-of-stack-pointer
## Severity: critical

`src/security.rs:363-460` (credential_callback_bridge), `693-768` (validation_callback_bridge).

OpenPMIx's `getcbfunc`/`valid_cbfunc` (pmix_security.c:101-109, 321-327) pass a **stack** `pmix_byte_object_t` and a PMIx-owned info array to the callback and destruct/free both immediately after it returns. The Rust bridges call `copy_and_free_pmix_byte_object` (libc `free` on PMIx's stack struct and its bytes) and `PMIx_Info_free(info)` / wrap info in `ValidationResults` whose Drop frees it — all double-free or free-of-stack-pointer. Additionally the "copy" loop shallow-copies `pmix_info_t` values then frees the original array, leaving dangling value pointers that `Info::drop` frees a second time.

**Fix:** never free in these bridges; deep-copy bytes/values (e.g. `PMIx_Info_xfer` into fresh `PMIx_Info_create` entries) and let PMIx release its own.

### PR DESCRIPTION
**Pull request by SedahsDev**
Closes #437

## Summary
- Stop non-blocking security callback bridges from freeing PMIx-owned credential and info memory.
- Deep-copy callback info payloads with PMIx_Info_xfer before exposing them to Rust-owned values.

## Affected functions
- credential_callback_bridge
- validation_callback_bridge
- copy_pmix_byte_object

## Rationale
PMIx supplies callback-owned byte-object and info storage that it releases after the callback returns. The previous bridges freed those objects and shallow-copied payload pointers, causing free-of-stack-pointer, double-free, and dangling-value bugs.

## Test plan
- Local mock-based security unit tests: cargo test --lib security::tests (67 passed, 2 ignored).
- cargo build with PMIX_PREFIX=/home/bzf/.local/openpmix-6.1.0 (passed).
- cargo fmt --check and git diff --check (passed for the change).
- Daemon-based tests are deferred to CI because no prterun/PMIx daemon is available locally.
- Full cargo test and cargo clippy --all-targets remain blocked by pre-existing repository test/lint failures unrelated to this change.

### CODE REVIEW

File: src/security.rs [hunk 1]
```diff
@@ -289,7 +297,7 @@ pub fn get_credential(info: &[Info]) -> Result<PmixCredential, PmixStatus> {
             // SAFETY: cred_ptr is valid. PMIx has populated it with
             // the credential bytes (allocated by pmix_malloc).
             // We copy the bytes into a Rust Vec and free the C memory.
-            let bytes = copy_and_free_pmix_byte_object(cred_ptr);
+            let bytes = copy_pmix_byte_object(cred_ptr);
```
Comment (SedahsReview) [inline src/security.rs:299]: **Warning - memory leak in the sync path.** This is the *sync* `PMIx_Get_credential`, not the async callback. Here `cred_ptr` is a Rust allocation (`Box::into_raw`, line 274) and `cred.bytes` was pmix_malloc'd by PMIx for this sync call. PMIx does NOT free these on a sync return - the "PMIx frees after the callback returns" guarantee from #437 applies only to the async callback bridges. Switching this line to copy-only means every successful sync `get_credential` now leaks both the Box struct and the pmix_malloc'd bytes. Note the error path below still correctly recovers the Box and frees the bytes (lines 308-310); the success path should do the same: copy the bytes, `ffi::free(cred.bytes)`, then `Box::from_raw(cred_ptr)` to release the struct. (The old `ffi::free(cred_ptr)` on the Box was allocator-mismatched UB; recovering the Box is the correct fix.)

File: src/security.rs [hunk 2]
```diff
@@ -199,27 +201,33 @@ impl PmixCredential {
 /// - `cred` must be a valid, non-null pointer to a `pmix_byte_object_t`
 ///   allocated by PMIx (i.e., via `pmix_malloc` / libc `malloc`).
 /// - The caller must ensure the struct is valid and not already freed.
-unsafe fn copy_and_free_pmix_byte_object(cred: *mut ffi::pmix_byte_object_t) -> Vec<u8> {
+unsafe fn copy_pmix_byte_object(cred: *const ffi::pmix_byte_object_t) -> Vec<u8> {
```
Comment (SedahsReview) [inline src/security.rs:198]: **Warning - misleading docs.** The `# Safety` block above this function still says it "free[s] the PMIx-allocated memory" / "free[s] the PMIx-allocated struct" (lines 190-203). The function no longer frees anything - for a memory-ownership fix this doc states the exact opposite of the actual behavior. Rewrite it to say the helper is copy-only and leaves both the struct and its bytes for the caller/PMIx to free. Also the doc block is duplicated (lines 190-198 and 199-203 are near-identical).

File: src/security.rs [hunk 3]
```diff
@@ -199,27 +201,33 @@ impl PmixCredential {
...
-        unsafe {
-            ffi::free(obj.bytes as *mut std::ffi::c_void);
-        }
     }
-    // Free the struct itself using libc free.
-    unsafe {
-        ffi::free(cred as *mut std::ffi::c_void);
+}
+
+unsafe fn copy_info_array(info: *const ffi::pmix_info_t, ninfo: usize) -> *mut ffi::pmix_info_t {
```
Comment (SedahsReview) [inline src/security.rs:216]: **Suggestion - missing `# Safety`.** `copy_info_array` is a new `unsafe fn` but has no `# Safety` doc, unlike `copy_pmix_byte_object`. Please document its contract (valid non-null `pmix_info_t[ninfo]`; the returned array is owned and must be released with `PMIx_Info_free`).

File: src/security.rs [hunk 4]
```diff
@@ -788,23 +788,24 @@ extern "C" fn validation_callback_bridge(
 
     // Build ValidationResults from the info array.
     let results = if !info.is_null() && ninfo > 0 {
+        let copied = unsafe { copy_info_array(info, ninfo) };
```
Comment (SedahsReview) [inline src/security.rs:794]: **Suggestion - silent data drop on copy failure.** When `copy_info_array` fails (e.g. a `PMIx_Info_xfer` error) it returns null, and the caller here silently builds an empty `ValidationResults` with no error surfaced. The credential bridge likewise `continue`s and drops the failing entry. Consider surfacing the copy failure rather than returning partial/empty results silently.

File: src/security.rs [hunk 5]
```diff
@@ -1145,33 +1154,33 @@ mod tests {
         let bo_ptr = Box::into_raw(bo);
 
         // Call the helper — it copies bytes and frees the C memory
-        let result = unsafe { copy_and_free_pmix_byte_object(bo_ptr) };
+        let result = unsafe { copy_pmix_byte_object(bo_ptr) };
```
Comment (SedahsReview) [inline src/security.rs:1164]: **Suggestion - stale test comment + leaked fixtures.** The comment above (line 1156) still reads "Call the helper - it copies bytes and frees the C memory", but the helper no longer frees. These renamed tests `Box::into_raw` and `libc::malloc` without ever freeing (the helper no longer does), so they leak. Update the comment and have the tests explicitly free the struct/bytes to keep the suite leak-clean.

**Review by SedahsReview — CHANGES_REQUESTED** [review]
Reviewed commit c50387d. The core fix is correct: the async credential/validation callback bridges no longer free PMIx-owned stack pmix_byte_object_t / info arrays, the info copy is now a true deep copy via PMIx_Info_xfer into fresh PMIx_Info_create entries, copy_pmix_byte_object is copy-only (no free), and validation_callback_bridge no longer wraps PMIx's raw info array in a ValidationResults whose Drop -> PMIx_Info_free would double-free. No leftover ffi::free / PMIx_Info_free on PMIx-owned callback memory. Build passes, all 67 security tests pass, security.rs is fmt-clean, and clippy shows only the 6 pre-existing bindgen missing_safety_doc errors in bindings.rs (unrelated to this PR).

Requesting changes for one warning-level regression plus a misleading doc comment; see inline comments.

1. [WARNING] Sync get_credential now leaks the credential on every successful call (line 300).
2. [WARNING] copy_pmix_byte_object doc still claims it frees PMIx memory (lines 190-203).
3. [SUGGESTION] copy_info_array has no # Safety doc (line 216).
4. [SUGGESTION] Copy failure silently drops results (line 791).
5. [SUGGESTION] Renamed tests still claim the helper frees memory and now leak their fixtures (line 1157).

**Comment by SedahsReview** [comment]
## Code Review Summary

**Verdict: Changes Requested** (2 warnings, 3 suggestions)

### 🔴 Critical
None.

### ⚠️ Warnings
- **src/security.rs:300** — Memory leak in the *sync* `get_credential` path. In the sync call `cred_ptr` is a Rust `Box::into_raw` allocation and `cred.bytes` was `pmix_malloc`'d by PMIx for the sync call. PMIx does not free these on a sync return (the "PMIx frees after the callback" guarantee from #437 applies only to the async callback bridges). Switching to copy-only leaks both the Box struct and the bytes on every successful sync call. Fix: copy bytes, `ffi::free(cred.bytes)`, then `Box::from_raw(cred_ptr)` (mirroring the error path at lines 308-310).
- **src/security.rs:190-203** — `copy_pmix_byte_object` doc still claims it "free[s] the PMIx-allocated memory/struct" and is duplicated; the function no longer frees anything. Misleading docs on a security/ownership fix.

### 💡 Suggestions
- **src/security.rs:216** — New `unsafe fn copy_info_array` has no `# Safety` doc (rest of file documents it on unsafe helpers).
- **src/security.rs:791** — When `copy_info_array` fails it returns null and validation info is silently dropped (empty `ValidationResults`, no error). Consider surfacing the failure.
- **src/security.rs:1157** — Renamed tests still say the helper "frees the C memory" and now leak their `Box::into_raw`/`libc::malloc` fixtures since the helper no longer frees.

### ✅ Looks Good
- Async `credential_callback_bridge` / `validation_callback_bridge` no longer free PMIx-owned stack `pmix_byte_object_t` / info arrays — the core #437 fix is correct.
- Info copy is now a true deep copy: `PMIx_Info_create(1)` + `PMIx_Info_xfer` per entry (credential) and `copy_info_array` (validation), so Rust-owned `Info`/`ValidationResults` entries own their payloads and their Drop is a valid single free; PMIx freeing the original after the callback is also valid (no shared/dangling pointers).
- `validation_callback_bridge` no longer wraps PMIx's raw array in a `ValidationResults` whose Drop would double-free.
- No leftover `ffi::free` / `PMIx_Info_free` on PMIx-owned callback memory (only the sync error path frees its own pmix_malloc'd bytes, correctly).

### Quality gates (PMIX_PREFIX=$HOME/.local/openpmix-6.1.0)
- `cargo build` — PASS
- `cargo test --lib security::tests -- --test-threads=1` — 67 passed, 0 failed, 2 ignored
- `cargo clippy --lib -- -D warnings` — only 6 pre-existing bindgen `missing_safety_doc` errors in `target/.../bindings.rs` (unrelated to this PR); no new findings on the PR's added lines
- `cargo fmt --check` — security.rs clean (0 diffs); repo-wide pre-existing diffs unaffected

---
*Reviewed by Hermes Agent (SedahsReview)*

File: src/security.rs [hunk 6]
```diff
@@ -382,7 +394,7 @@ static CREDENTIAL_INFO_REGISTRY: LazyLock<Mutex<HashMap<usize, RetainedCredentia
 ///
 /// Ownership notes:
 /// - `credential` is allocated by PMIx — we copy its bytes into a Vec<u8>
-///   and free the PMIx-allocated memory.
+///   without freeing the callback-owned memory.
```
Comment (SedahsReview) [inline src/security.rs:397]: 💡 **Suggestion (stale doc — lines 398-399).** This ownership note still reads "`info` is allocated by PMIx — we copy the entries into a Vec<Info> and then free the original array via `PMIx_Info_free`." That free no longer happens (and must not — PMIx owns and frees this array after the callback). Please update the note to state we deep-copy without freeing, matching the credential bullet above and the fix's intent.

File: src/security.rs [hunk 7]
```diff
@@ -463,16 +465,18 @@ extern "C" fn credential_callback_bridge(
...
-                    std::ptr::copy_nonoverlapping(&entry as *const ffi::pmix_info_t, new_info, 1);
+                    if ffi::PMIx_Info_xfer(new_info, &entry) != ffi::PMIX_SUCCESS as i32 {
+                        ffi::PMIx_Info_free(new_info, 1);
+                        continue;
+                    }
                     info_vec.push(Info {
                         handle: new_info,
                         len: 1,
-                    _not_thread_safe: std::marker::PhantomData,
+                        _not_thread_safe: std::marker::PhantomData,
```
Comment (SedahsReview) [inline src/security.rs:475]: 💡 **Suggestion (dangling comment — line 479).** The comment `// Free the original C-allocated info array.` above now describes an operation that no longer occurs — the `PMIx_Info_free(info, ninfo)` call was removed, so line 479 sits directly above `let n = info_vec.len();`. This comment is exactly the anti-pattern this PR removes (freeing PMIx-owned callback memory) and could mislead a future edit into reintroducing a use-after-free. Please delete or correct it.

File: src/security.rs [hunk 8]
```diff
@@ -1553,11 +1579,11 @@ mod tests {
         }
     }
 
-    // ── copy_and_free_pmix_byte_object edge cases ─────────────────────────
+    // ── copy_pmix_byte_object edge cases ─────────────────────────
 
     /// Test copy_and_free with large data block.
     #[test]
-    fn test_copy_and_free_pmix_byte_object_large() {
+    fn test_copy_pmix_byte_object_large() {
```
Comment (SedahsReview) [inline src/security.rs:1586]: 💡 **Suggestion (leaked test fixtures).** Three `copy_pmix_byte_object` tests still leak their fixtures after the helper stopped freeing: `test_copy_pmix_byte_object_large` (1586), `test_copy_pmix_byte_object_with_nulls` (1605), and `test_copy_pmix_byte_object_single_byte` (1624) all `libc::malloc` the bytes buffer and `Box::into_raw` the struct but never reclaim them (the allocated/`null_bytes`/`zero_size` tests were given `ffi::free`+`Box::from_raw` cleanup, but these three were not). Please add the same cleanup so the fixtures don't leak.

**Review by SedahsReview — APPROVED** [review]
## Hermes Agent Re-Review — APPROVE

Re-review of the updated head (4cd550c) against prior review 4912939800. The two blocking warnings are resolved and the core security fix holds.

**Prior blocking issues — verified fixed:**
- **Sync `get_credential` success path leak** (src/security.rs:299-304): now copies the bytes, frees the PMIx-`pmix_malloc`'d `cred.bytes` via `ffi::free`, and reclaims the struct with `Box::from_raw(cred_ptr)` — correctly mirroring the error path. PMIx does not free on the sync return, so the caller must; this is now done exactly once, no leak and no double-free.
- **`copy_pmix_byte_object` doc** (191-197): now accurately states it only copies the bytes and does **not** free the struct or its storage; ownership remains with the caller/PMIx.

**Core fix confirmed:**
- No double-free, no free-of-stack-pointer anywhere in the diff.
- Async `credential_callback_bridge` / `validation_callback_bridge` are copy-only on PMIx-owned callback memory — PMIx frees it after the callback returns.
- Callback info payloads are deep-copied via `PMIx_Info_create`+`PMIx_Info_xfer` (credential) or `copy_info_array` (validation); deep-copied entries are freed exactly once by their own `Info::drop`/`ValidationResults::drop`. The original PMIx arrays are left for PMIx.
- `copy_info_array` (209-231) now carries a `# Safety` doc documenting the `PMIx_Info_free` ownership contract, and a copy failure is surfaced as `ErrNomem` (line 798) instead of silently dropping data.
- Both new `unsafe fn`s carry `# Safety` sections.

**Quality gates:** `cargo build` ✓ · `cargo test --lib security::tests -- --test-threads=1` 67 passed / 2 ignored ✓ · `cargo clippy --lib -- -D warnings` — only pre-existing bindgen `missing_safety_doc` errors in generated bindings, none in security.rs ✓ · `cargo fmt --check` — security.rs clean (only pre-existing examples diffs) ✓

**Non-blocking suggestions** (3 inline comments — please tidy before merge): stale ownership doc note at 398-399, dangling 'free the C array' comment at 479, and three `copy_pmix_byte_object` tests that still leak their malloc/Box fixtures.

**Comment by SedahsReview** [comment]
## Code Review Summary — VERDICT: APPROVE ✅

Re-review of updated head 4cd550c (2 commits) for PR #458. The two blocking WARNINGs from prior review 4912939800 are resolved and the core security fix is confirmed correct.

### Critical — none

### Warnings (prior, now fixed)
- **Sync get_credential success path leak** (security.rs:299-304): copies bytes, frees PMIx-malloc'd `cred.bytes` via `ffi::free`, reclaims struct via `Box::from_raw` — mirrors error path. No leak, no double-free.
- **copy_pmix_byte_object doc** (191-197): now correctly states it only copies bytes and does not free the struct or its byte storage.

### Core fix holds
- No double-free / free-of-stack-pointer anywhere.
- Async credential and validation bridges are copy-only on PMIx-owned callback memory (PMIx frees it after the callback returns).
- Callback info deep-copied via PMIx_Info_create + PMIx_Info_xfer / copy_info_array; deep-copied entries freed exactly once by their own Info::drop / ValidationResults::drop; original PMIx arrays left to PMIx.
- copy_info_array has a # Safety doc for the PMIx_Info_free ownership contract; a copy failure is surfaced as ErrNomem (line 798) instead of silently dropping data.
- Both new unsafe fns have # Safety sections.

### Suggestions (non-blocking, see inline comments)
- Stale ownership doc note at 398-399 claims we free the info array via PMIx_Info_free (no longer true).
- Dangling comment at 479 "Free the original C-allocated info array." describes a removed operation.
- 3 tests (large 1586 / with_nulls 1605 / single_byte 1624) still leak their malloc/Box fixtures.

### Quality gates
cargo build ✓ · security tests 67 passed / 2 ignored ✓ · clippy — only pre-existing bindgen missing_safety_doc in generated bindings, none in security.rs ✓ · cargo fmt --check — security.rs clean ✓

Formal review: https://github.com/SedahsDev/pmix-rs/pull/458#pullrequestreview-4912987166

---
*Reviewed by Hermes Agent (code-reviewer profile)*
